File: src/main/java/io/github/yuokada/simulations/LoadSimulationConfig.java
package io.github.yuokada.simulations;

import java.io.IOException;
import java.io.InputStream;
import java.util.Properties;

import static io.github.yuokada.simulations.SimulationConstants.DEFAULT_API_ENDPOINT;
import static io.github.yuokada.simulations.SimulationConstants.SIMULATION_PREFIX;

public class LoadSimulationConfig {

    public static TdApiConfig loadApiConfig() {
        Properties props = getProperties();

        String endpoint = props.getProperty(SIMULATION_PREFIX + "endpoint", DEFAULT_API_ENDPOINT);
        String token = props.getProperty(SIMULATION_PREFIX + "api-token", "");
        if (System.getenv("TD_API_ENDPOINT") != null) {
            endpoint = System.getenv("TD_API_ENDPOINT");
        }
        if (System.getenv("TD_API_KEY") != null) {
            token = System.getenv("TD_API_KEY");
        }
        return new TdApiConfig(endpoint, token);
    }

    public static SimulationConfig loadSimulationConfig() {
        Properties props = getProperties();
        String schema = props.getProperty(SIMULATION_PREFIX + "schema", null);
        String maxClients = props.getProperty(SIMULATION_PREFIX + "max-concurrent-clients", String.valueOf(3));
        String testDuration = props.getProperty(SIMULATION_PREFIX + "test-duration-seconds", String.valueOf(60));
        return new SimulationConfig(schema, Integer.valueOf(maxClients), Integer.valueOf(testDuration));
    }

    private static Properties getProperties() {
        // Load configuration from src/main/resources/application.properties
        Properties props = new Properties();
        try (InputStream input = LoadSimulationConfig.class.getClassLoader()
            .getResourceAsStream("application.properties")) {
            if (input == null) {
                throw new IllegalStateException("Unable to find application.properties");
            }
            props.load(input);
        } catch (IOException e) {
            throw new RuntimeException("Failed to load configuration", e);
        }
        return props;
    }
}

File: src/test/scala/io/github/yuokada/simulations/LoadSimulationConfig.scala
package io.github.yuokada.simulations

import java.io.{IOException, InputStream}
import java.util.Properties

object LoadSimulationConfig {

  def loadApiConfig(): TdApiConfig = {
    val props = getProperties()

    var endpoint =
      props.getProperty(SimulationConstants.SIMULATION_PREFIX + "endpoint", SimulationConstants.DEFAULT_API_ENDPOINT)
    var token = props.getProperty(SimulationConstants.SIMULATION_PREFIX + "api-token", "")

    if (sys.env.contains("TD_API_ENDPOINT")) {
      endpoint = sys.env("TD_API_ENDPOINT")
    }
    if (sys.env.contains("TD_API_KEY")) {
      token = sys.env("TD_API_KEY")
    }

    TdApiConfig(endpoint, token)
  }

  def loadSimulationConfig(): SimulationConfig = {
    val props = getProperties()

    val schema       = Option(props.getProperty(SimulationConstants.SIMULATION_PREFIX + "schema")).getOrElse(null)
    val maxClients   = props.getProperty(SimulationConstants.SIMULATION_PREFIX + "max-concurrent-clients", "3").toInt
    val testDuration = props.getProperty(SimulationConstants.SIMULATION_PREFIX + "test-duration-seconds", "60").toInt

    SimulationConfig(schema, maxClients, testDuration)
  }

  private def getProperties(): Properties = {
    val props        = new Properties()
    val resourcePath = "application.properties"

    val input: InputStream = Option(getClass.getClassLoader.getResourceAsStream(resourcePath))
      .getOrElse(throw new IllegalStateException(s"Unable to find $resourcePath"))

    try {
      props.load(input)
    } catch {
      case e: IOException => throw new RuntimeException("Failed to load configuration", e)
    } finally {
      input.close()
    }

    props
  }
}
